persistence/sql/process_local_queue.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 XDBLab organization
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"github.com/xdblab/xdb-apis/goapi/xdbapi"
+	"github.com/xdblab/xdb/common/log/tag"
+	"github.com/xdblab/xdb/extensions"
+	"github.com/xdblab/xdb/persistence"
+)
+
+func (p sqlProcessStoreImpl) ProcessLocalQueueMessage(
+	ctx context.Context, request persistence.ProcessLocalQueueMessageRequest,
+) error {
+	tx, err := p.session.StartTransaction(ctx, defaultTxOpts)
+	if err != nil {
+		return err
+	}
+
+	err = p.doProcessLocalQueueMessageTx(ctx, tx, request)
+	if err != nil {
+		err2 := tx.Rollback()
+		if err2 != nil {
+			p.logger.Error("error on rollback transaction", tag.Error(err2))
+		}
+	} else {
+		err = tx.Commit()
+		if err != nil {
+			p.logger.Error("error on committing transaction", tag.Error(err))
+			return err
+		}
+	}
+	return err
+}
+
+func (p sqlProcessStoreImpl) doProcessLocalQueueMessageTx(
+	ctx context.Context, tx extensions.SQLTransaction, request persistence.ProcessLocalQueueMessageRequest,
+) error {
+	// Step 1: update process execution row
+	prcRow, err := tx.SelectProcessExecutionForUpdate(ctx, request.ProcessExecutionId)
+	if err != nil {
+		return err
+	}
+
+	waitingQueues, err := persistence.NewStateExecutionWaitingQueuesFromBytes(prcRow.StateExecutionWaitingQueues)
+	if err != nil {
+		return err
+	}
+
+	dedupIdString := request.DedupId.String()
+	assignedStateExecutionId, hasFinishedWaiting := waitingQueues.Consume(xdbapi.LocalQueueMessage{
+		QueueName: request.QueueName,
+		DedupId:   &dedupIdString,
+		Payload:   &request.Payload,
+	})
+
+	// early stop if no state can consume the message
+	if assignedStateExecutionId == nil {
```

Comment:
What if the message can be consumed later? 
A message can be sent before the waitUntil API is completed, or there can be a different async state interested in the message later. 

In additional to what states are interested in what queues, we probably also need to store what are the remaining messages in processExecution for this feature. 

Also another question -- what if the command is waiting for more than one message for the same queue? (maybe we have to move the message from the queue into commandResults)

Reply:
Oh I missed this part. Now I have added `UnconsumedMessages` and implemented the logic to deal with it.

> what if the command is waiting for more than one message for the same queue?
This logic has been implemented previously in this PR. The related logic is in `func (s *StateExecutionWaitingQueuesJson) Consume(message LocalQueueMessageInfoJson) (*string, bool)`.